File: Dsa-day196.py
class Solution:
    def minimumCost(self, source: str, target: str, original: List[char], changed: List[char], cost: List[int]) -> int:
        dist = [[float('inf')] * 26 for _ in range(26)]
        
        for i in range(26):
            dist[i][i] = 0
        
        for i in range(len(cost)):
            u = ord(original[i]) - ord('a')
            v = ord(changed[i]) - ord('a')
            dist[u][v] = min(dist[u][v], cost[i])
        
        for k in range(26):
            for i in range(26):
                for j in range(26):
                    if dist[i][k] != float('inf') and dist[k][j] != float('inf'):
                        dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
        
        totalCost = 0
        
        for i in range(len(source)):
            if source[i] != target[i]:
                u = ord(source[i]) - ord('a')
                v = ord(target[i]) - ord('a')
                
                if dist[u][v] == float('inf'):
                    return -1
                
                totalCost += dist[u][v]
        
        return totalCost

File: Dsa-day196.java
class Solution {
    public long minimumCost(String source, String target, char[] original, char[] changed, int[] cost) {
        long[][] dist = new long[26][26];
        
        for(int i = 0; i < 26; i++) {
            for(int j = 0; j < 26; j++) {
                if(i == j) {
                    dist[i][j] = 0;
                } else {
                    dist[i][j] = Long.MAX_VALUE;
                }
            }
        }
        
        for(int i = 0; i < cost.length; i++) {
            int u = original[i] - 'a';
            int v = changed[i] - 'a';
            dist[u][v] = Math.min(dist[u][v], cost[i]);
        }
        
        for(int k = 0; k < 26; k++) {
            for(int i = 0; i < 26; i++) {
                for(int j = 0; j < 26; j++) {
                    if(dist[i][k] != Long.MAX_VALUE && dist[k][j] != Long.MAX_VALUE) {
                        dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                    }
                }
            }
        }
        
        long totalCost = 0;
        
        for(int i = 0; i < source.length(); i++) {
            if(source.charAt(i) != target.charAt(i)) {
                int u = source.charAt(i) - 'a';
                int v = target.charAt(i) - 'a';
                
                if(dist[u][v] == Long.MAX_VALUE) {
                    return -1;
                }
                
                totalCost += dist[u][v];
            }
        }
        
        return totalCost;
    }
}